### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by risolate**
어느 정도 완성이 된 것 같아 main으로의 pr을 제안합니다

### CODE REVIEW

File: generation/reward_util.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        e = e[0] if 0<len(e) and 4<e[0]<i else i-1
+        tgt.append(s[:e].cpu().tolist())
+    tgt_idx = sc_tokenizer(
+        tokenizer.batch_decode(tgt),
+        max_length=128,
+        padding="max_length",
+        truncation=True,
+        return_tensors="pt",
+    )["input_ids"].to(device)
+    tgt_cls = classifier(tgt_idx).detach()
```
Comment (intrandom5) [inline generation/reward_util.py:72]: classification 모델로 inference만 하면 되는 거죠?
그러면 inference한 뒤에 detach()를 쓰는 것보다 with torch.no_grad(): 를 사용하는게 자원적으로 더 효율적일 것 같습니다.

File: generation/train.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+from torch.utils.data import Dataset, DataLoader
+from transformers import (
+    AutoTokenizer,
+    MBartForConditionalGeneration,
+    is_torch_available,
+    T5ForConditionalGeneration,
+    T5Tokenizer
+
+)
+sys.path.append("..")
```
Comment (intrandom5) [inline generation/train.py:16]: **이 부분은 지극히 제 개인적인 생각인데,**
classification 폴더의 CNNModel을 불러오기 위해서 추가된 것 같은데 CNN Model 클래스를 generation 폴더의 모델 폴더를 만들어서 추가하는 게 더 좋지 않을까? 하는 생각이 들었습니다.
generation 폴더와 classification 폴더는 같은 repo에 있지만 서로 다른 모듈인데 이런 코드를 짜는 것이 맞나? 하는 생각이 조금 들어서. 이 부분은 잘 모르겠네요 다른 repo들은 어떤 식으로 하는지

File: generation/reward_util.py (hunk 1 above)
Comment (intrandom5) [reply generation/reward_util.py:72]: ## 코드
그 외에는 뭐 딱히 제 편협한 시각으로는 코드 동작 상의 흠이 있는 것 같진 않습니다.
## 실험
### 1. bl_loss
실험적으로는, bl_loss가 sc_loss나 ce_loss에 비해 현저히 작습니다. 이러면 bl_loss는 있으나마나 한게 아닌가 싶네요.
본래 style transfer를 수행하는 모델이기 때문에 source와 target 텍스트가 거의 유사한 경우를 학습하여 bl_loss가 쓰였을 것 같습니다.
하지만 사실 수집한 데이터들을 보면 bl_loss를 통해 '두 문장 사이에 단어가 얼마나 겹치는가?'를 평가하는 것은 오히려 학습에 방해가 될 수도 있을 것 같습니다.

### 2. 다른 classifier 모델 사용
그리고 sc_loss에 다른 Classifier 모델들을 사용해 보는 방법도 있을 것 같습니다! 지금은 수현이가 준 CNN 모델로 수행하고 있는 것 같은데, 다른 classifier 모델들을 다르게 사용해 볼 수도 있겠죠?
sc_loss에서는 classifier의 recall보다 precision(혐오를 혐오로 분류하는 지표)이 더 중요한 것 같습니다. 따라서 서비스에 사용할 k-mhas 데이터로 훈련된 모델보다 precision이 높은 다른 데이터셋으로 훈련된 모델을 사용하는 것이 좋을 것 같습니다. (이건 저도 생각해 보겠습니다)
그리고 서비스 하는게 아니니까 inference 속도도 관계 없겠죠? (학습이 조금 느려지면 어떻습니까) transformer 기반의 KcELECTRA-base-v2022 기반의 모델로 바꿔 보는 것도 좋을 것 같습니다. (제가 훈련시켜서 줘야 할 것 같지만...)

### 3. Threshold 적용
이와 비슷하게 classifier의 threshold 같은 것도 설정할 수 있다면 (코드 상에서는 보이지 않네요. 제가 reward loss 계산을 잘 몰라서 모르겠습니다) 이를 조금 낮춰보면 (혐오 표현으로 분류하는 기준을 낮춘다면) 모델이 좀 더 혐오 표현, 단어에 대해서 민감하게 학습될 수 있지 않을까? 하는 생각도 들었습니다!

이상입니다. 확인 되면 알아서 merge 눌러 주시면 될 것 같습니다. 재환이형 짱!

Comment (risolate) [reply generation/reward_util.py:72]: 유민이 짱

Comment (ymnseol) [reply generation/reward_util.py:72]: BLEU의 경우 source가 아닌 target text를 기준으로 계산되도록 되어있습니다! 그래서 '실제로 사람이 순화한 텍스트와, 모델이 생성한 순화 텍스트가 얼마나 유사한지'를 판단합니다. 순화 시 문장이 source와 많이 달라지더라도 문제가 없을 것으로 판단됩니다! 다만 BLEU loss가 작아서 거의 영향력을 못 미친다는 점은 충분히 고려해 볼 부분인 것 같습니다. 사람이 직접 순화한 데이터를 쓰기 때문에, 데이터의 특성을 잘 반영하도록 모델을 학습시킬 수 있을 것 같네요! 좋은 인사이트 공유 감사합니다😀
